kaldb/src/main/java/com/slack/kaldb/chunk/manager/caching/CachingChunkManager.java

```diff
@@ -0,0 +1,68 @@
+package com.slack.kaldb.chunk.manager.caching;
+
+import static com.slack.kaldb.config.KaldbConfig.DEFAULT_START_STOP_DURATION;
+
+import com.slack.kaldb.chunk.ReadOnlyChunkImpl;
+import com.slack.kaldb.chunk.manager.ChunkManager;
+import com.slack.kaldb.logstore.LogMessage;
+import com.slack.kaldb.proto.config.KaldbConfigs;
+import com.slack.kaldb.server.MetadataStoreService;
+import io.micrometer.core.instrument.MeterRegistry;
+import java.io.IOException;
+import java.util.UUID;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CachingChunkManager<T> extends ChunkManager<T> {
+  private static final Logger LOG = LoggerFactory.getLogger(CachingChunkManager.class);
+
+  private final MeterRegistry meterRegistry;
+  private final MetadataStoreService metadataStoreService;
+  private final KaldbConfigs.CacheConfig cacheConfig;
+
+  public CachingChunkManager(
+      MeterRegistry registry,
+      MetadataStoreService metadataStoreService,
+      KaldbConfigs.CacheConfig cacheConfig) {
+    this.meterRegistry = registry;
+    this.metadataStoreService = metadataStoreService;
+    this.cacheConfig = cacheConfig;
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    LOG.info("Starting caching chunk manager");
+    metadataStoreService.awaitRunning(DEFAULT_START_STOP_DURATION);
+
+    for (int i = 0; i < cacheConfig.getSlotsPerInstance(); i++) {
```

Comment:
I ended up refactoring this from the original [IntStream on the prior PR](https://github.com/slackhq/kaldb/pull/117/files#diff-0e217760e60aa380caf8feb80e8ab69c44231b5d36a6ebe627fe7d6822d5ee46R57), as the `ReadOnlyChunkImpl` can throw an exception on instantiation. In this specific case we likely would prefer to fail startup of the application instead of allowing some number of slots less than our target to instantiate, as this would never be retried.